### PR TITLE
🎉 Source Bing Ads: implement `OAuth2.0` support, remove `redirect_uri`, change `Account ID` to `User ID` in spec

### LINKED ISSUE #6294: Source Bing: support oauth
## Tell us about the problem you're trying to solve
With the release of Airbyte Cloud, we need to start supporting Oauth for this connector, since it's the recommended way of authenticating users into a SaaS application. 

If this connector doesn't support oauth already (i.e: doesn't accept a client_id and client_secret) then we need to update its spec to accept those parameters. I suggest that this be a oneof nested inside a top-level field called "authentication": 

`
{
  authentication: {
     type: object
     oneOf: [ 
       // api key,
       // oauth
     ] 
  }
}
`

See the connector spec reference in the docs for reference on how a oneof can be implemented.

This should be done in a backwards compatible manner i.e: users currently supplying authentication info in the config's top-level should not be impacted by this change. 

## Acceptance Criteria

1. The connector supports oauth webflow authentication with client_id/client_secret 
2. Oauth properties are annotated properly. See [this PR](https://github.com/airbytehq/airbyte/pull/6022) for an example

### LINKED ISSUE #9972: Source Bing Ads: change account id field name
I tried to setup the Bing Ads connector as source. However following the exact steps I got the following error: `"Exception(\"You don't have accounts assigned to this user.\")"` . 

So I followed the whole setup again step by step and Airbyte requested an Account ID from me. When I went to the docs Microsoft said:
> _The URL will contain a cid key/value pair in the query string that identifies your customer ID, and an aid key/value pair that identifies your account ID._

So I thought the Account Id was the aid key/value pair. However: when I inserted it in Airbyte it the gave above error. Well, cue 3 hours later. I thought let's have a look at the JSON  request/response of airbyte. And then I saw the following JSON.

```
{
   "sourceDefinitionId":"xxxxxx-xxxxxxx-xxxxxx-xxxxxx",
   "connectionConfiguration":{
      "reports_start_date":"2020-01-01",
      "monthly_reports":false,
      "developer_token":"xxxxxxxxxxxxxxxx",
      "weekly_reports":false,
      "hourly_reports":false,
      "refresh_token": "xxx....xxxxxxx",
      "daily_reports":true,
      "client_secret":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "customer_id":"xxxxxxx",
      "client_id":"xxxxxx-xxxxxx-xxxxxx-xxxxxx",
      "accounts":{
         "selection_strategy":"all"
      },
      "user_id":"12345678"
   }
}
```
So I found out that the field that in Airbyte, which is called Account Id, was send as user_id . On top of that, in the source of the Bing Ads connector I found the following: [https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L97](https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L97). 

So I dove into it further, and checked the url which microsoft provided. I found three values: cid, aid and **uid**
https://ads.microsoft.com/cc/Users?cid=12345678&aid=12345678&uid=12345678987 . 
Well, what are the chances that this is the user_id that is being send over JSON. I swapped out the values and BAM, miracle happened. It worked!

So I forked the repo and tried to change the value, but I couldn't test it. So I did my best to document the problem and point you all in the right direction.

TLDR
Account Id in the Bing ads connector should be called User Id. The file/line where it should be changed is 

[https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L97](https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L97)
&&
[https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L98](https://github.com/airbytehq/airbyte/blob/98b5aabd187a7338e01114c373a397d3aae3590e/airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json#L98). 

Account Id should be changed to User Id

### LINKED ISSUE #12488: Source Bing Ads: redirect_url is not needed
## Current Behavior
The redirect_url parameter requested by the connector is not needed. You'll notice that the connector only uses the method `request_oauth_tokens_by_refresh_tokens` which does not require the use of redirect_url. Therefore it's not required for the parameter to be included in spec.json and should be removed. 

## Expected Behavior
remove the parameter from spec.json in a backwards compatible way (keeping additionalProperties=true so as not to break the config for existing users)

### LINKED ISSUE #12490: Source Bing Ads: remove the user_id parameter from spec.json
## Current Behavior
The user_id parameter is not needed given that we can use the [`getuser` endpoint](https://docs.microsoft.com/en-us/advertising/customer-management-service/getuser?view=bingads-13) to get it on behalf of the user which performed the oauth process

## Expected Behavior
call [getuser](https://docs.microsoft.com/en-us/advertising/customer-management-service/getuser?view=bingads-13) to get the user id instead of requiring it from the user

In fact, we can remove all the parameters required here: user_id, account_id, and campaign_id by following the steps provided [here](https://docs.microsoft.com/en-us/advertising/guides/get-started?view=bingads-13#get-ids) to get all those parameters for the given authenticated user.

### LINKED ISSUE #12925: Source Bing Ads: should have at least 90% unit test coverage (currently 64%)

### LINKED ISSUE #12939: 🎉 Source Bing Ads: label `Bing Ads` as `beta` in the UI

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving:
https://github.com/airbytehq/airbyte/issues/9972
https://github.com/airbytehq/airbyte/issues/6294
https://github.com/airbytehq/airbyte/issues/12488
https://github.com/airbytehq/airbyte/issues/12939
https://github.com/airbytehq/airbyte/issues/12490
https://github.com/airbytehq/airbyte/issues/12925

## How
- edited `spec.json` to reflect oauth support
- added support of old input configuration for easy connector update transition for Customers
- edited `client.py` to support oauth config
- added `MicrosoftBingAdsOAuthFlow.java` class - oauth implementation
- edited `spec.json` to get rid of `accounts`, `user_id`, `customer_id` from input configuration, get these values from authenticated user instead.
- increased `unit_tests` coverage up to `92%`

## 🚨 User Impact 🚨
No impact expected.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

### CODE REVIEW

**Review by lazebnyi — APPROVED** [review]
LGTM

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml [hunk 1]
```diff
@@ -108,7 +108,7 @@
   documentationUrl: https://docs.airbyte.io/integrations/sources/bing-ads
   icon: bingads.svg
   sourceType: api
-  releaseStage: alpha
+  releaseStage: beta
```
Comment (evantahler) [inline airbyte-config/init/src/main/resources/seed/source_definitions.yaml:111]: 🥳

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/schemas/accounts.json [hunk 2]
```diff
@@ -80,7 +80,7 @@
       "type": ["null", "number"]
     },
     "PauseReason": {
-      "type": ["null", "string"]
+      "type": ["null", "number"]
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/schemas/accounts.json:83]: Oh Microsoft - looks like this is a number (unsignedByte) that can be used to look up a reason later - https://docs.microsoft.com/en-us/advertising/customer-management-service/advertiseraccount?view=bingads-13

Good catch!

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json [hunk 3]
```diff
@@ -5,125 +5,125 @@
...
-              "selection_strategy": {
+              "auth_method": {
+                "type": "string",
+                "const": "private_client",
+                "order": 0
+              },
+              "developer_token": {
+                "type": "string",
+                "title": "Developer Token",
+                "description": "Developer token associated with user. See more info <a href=\"https://docs.microsoft.com/en-us/advertising/guides/get-started?view=bingads-13#get-developer-token\"> in the docs</a>.",
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:91]: Can you please confirm that we do render HTML properly in the UI? I don't know if we sanitize those strings or not

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/MicrosoftBingAdsOAuthFlow.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import io.airbyte.oauth.BaseOAuth2Flow;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Supplier;
+import org.apache.http.client.utils.URIBuilder;
+
+public class MicrosoftBingAdsOAuthFlow extends BaseOAuth2Flow {
```
Comment (evantahler) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/MicrosoftBingAdsOAuthFlow.java:19]: Nice!

I think that
1) This new class needs to be registered in the factory - https://github.com/airbytehq/airbyte/blob/master/airbyte-oauth/src/main/java/io/airbyte/oauth/OAuthImplementationFactory.java#L22
2) Please add a test like https://github.com/airbytehq/airbyte/blob/master/airbyte-oauth/src/test/java/io/airbyte/oauth/flows/DriftOAuthFlowTest.java

**Review by evantahler — CHANGES_REQUESTED** [review]
This is my first time reviewing an oAuth addition, so I'll add @girarda as a reviewer as well

File: airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml [hunk 5]
```diff
@@ -3,6 +3,8 @@ tests:
   spec:
     - spec_path: "source_bing_ads/spec.json"
   connection:
+    - config_path: "secrets/config_old.json"
```
Comment (girarda) [inline airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml:6]: can this extra config_path be deleted?

File: airbyte-integrations/connectors/source-bing-ads/unit_tests/test_source.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def test_accounts_live(config):
+    client = Client(**config)
+    accounts = Accounts(client, config)
+    records = accounts.read_records(SyncMode.full_refresh)
+    assert len(list(records)) == 4
+
+
+# def test_AccountPerformanceReportMonthly_live(config):
```
Comment (girarda) [inline airbyte-integrations/connectors/source-bing-ads/unit_tests/test_source.py:151]: can you delete the dead code?

File: airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml (hunk 5 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml:6]: This is a needed part for testing the support of old input configurations, is there any valid reason we need to remove it?

File: airbyte-integrations/connectors/source-bing-ads/unit_tests/test_source.py (hunk 6 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-bing-ads/unit_tests/test_source.py:151]: Done.

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json (hunk 3 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:91]: Screenshot 2022-05-21 at 22 18 06

It shown something like in this example. The connector on the screenshot is different, but the concept is the same.

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/MicrosoftBingAdsOAuthFlow.java (hunk 4 above)
Comment (bazarnov) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/MicrosoftBingAdsOAuthFlow.java:19]: 1. The factory has it registered.
2. Done.

File: airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml (hunk 5 above)
Comment (girarda) [reply airbyte-integrations/connectors/source-bing-ads/acceptance-test-config.yml:6]: makes sense!

**Review by girarda — APPROVED** [review]

**Review by evantahler — APPROVED** [review]
